typescript/sdk/src/core/app.ts

```diff
@@ -1,37 +1,56 @@
-import { Inbox, InboxValidatorManager } from '@abacus-network/core';
-
 import { AbacusApp } from '../app';
-import { domains } from '../domains';
-import { ChainName, NameOrDomain } from '../types';
-
-import { CoreContractAddresses, CoreContracts } from './contracts';
+import { MultiProvider } from '../provider';
+import { ChainName, Remotes } from '../types';
 
-export class AbacusCore extends AbacusApp<
+import {
   CoreContractAddresses,
-  CoreContracts
-> {
-  constructor(addresses: Partial<Record<ChainName, CoreContractAddresses>>) {
-    super();
-    const chains = Object.keys(addresses) as ChainName[];
-    chains.map((chain) => {
-      this.registerDomain(domains[chain]);
-      const domain = this.resolveDomain(chain);
-      this.contracts.set(domain, new CoreContracts(addresses[chain]!));
-    });
+  CoreContractSchema,
+  CoreContracts,
+} from './contracts';
+import { environments } from './environments';
+
+type Environments = typeof environments;
+type EnvironmentName = keyof Environments;
+
+export class AbacusCore<
+  Networks extends ChainName = ChainName,
+> extends AbacusApp<CoreContracts<Networks>, Networks> {
+  constructor(
+    networkAddresses: {
+      [local in Networks]: CoreContractAddresses<Networks, local>;
+    },
+    multiProvider: MultiProvider<Networks>,
+  ) {
+    super(CoreContracts, networkAddresses, multiProvider);
+  }
+
+  static fromEnvironment(
+    name: EnvironmentName,
+    multiProvider: MultiProvider<any>,
+  ) {
+    return new AbacusCore(environments[name], multiProvider);
+  }
+
+  // override type to be derived from network key
+  getContracts<Local extends Networks>(
+    network: Local,
+  ): CoreContractSchema<Networks, Local> {
+    return super.getContracts(network) as any;
```

Comment:
Why the as any here? Isn't the return type well defined?

Reply:
no, the `<Networks, Local>` is not well defined